frame/contracts/src/exec.rs

```diff
@@ -44,6 +44,15 @@ bitflags! {
 	}
 }
 
+/// Describes where a call originated.
+pub enum CallOrigin {
```

Comment:
I think that might be a bit confusing since so far there is a noun Origin that is used to describe the top level caller. I.e. we can say that `ExternallyOwned` == `origin`. Here, again using the existing terminology, we can refer to this as the "type" of the sender (aka the caller) account.

Reply:
Writing this I also realized, that at some point if we introduce back `ext_dispatch_call` or some how introduce an ability for a contract to call back into the contract what we would really want to have here is the dichotomy: Contract / PlainAccount.

And in case of a plain account we use the existential deposit as the floor. If it is a contract we use subsistence threshold as the floor.

Reply:
So therefore, how about this:

1. we rename this enum struct to `TransactorKind`.
2. give it two variants: `Contract` and `PlainAccount` (i.e. an account without contract).
3. for now, we can leave this heuristic that if the `depth = 0` then we can say that it is always `PlainAccount`, but as a courtesy for future-maintainers (likely ourselves) we leave a debug assert that states that `ContractInfo` of the transactor at depth 0 is `None`

Reply:
I agree with your proposal. I think in the future we will have two different types of `ExecutionContext` that will improve this check ( #6448 ).